File: DesktopSample/DataBinding/Section/IListBinding/C#/IListBinding.DataLayer/ProductCollection.cs
using System;
using System.ComponentModel;
using System.Collections;
using System.Data;
using System.Data.SQLite;

namespace ActiveReports.Samples.IListBinding.DataLayer
{
	public class ProductCollection : CollectionBase, IComponent
	{
		private ISite _site;  //IComponentの実装に必要です。
		public ProductCollection()
		{
			FillWithAllProducts();
			_site = null;
		}
	
		public void FillWithAllProducts()
		{
			// 現在のリストをリセットします。
			List.Clear();
			
			using (var cn = DataProvider.NewConnection)
			{
				cn.Open();
				var cmd = new SQLiteCommand("SELECT * FROM Products", cn);
				IDataReader reader = cmd.ExecuteReader();

				while (reader.Read())
				{
					List.Add(ProductFromDataReader(reader));
				}
			}
		}

		/// <summary>
		/// ProductFromDataReader: リストのデータベースから新しい製品を作成するために使用します。
		/// </summary>
		/// <param name="reader">データを含む有効なDataReaderオブジェクト。</param>
		/// <returns>データ行から新しく作成されたProductオブジェクト。</returns>
		internal static Product ProductFromDataReader(IDataReader reader)
		{
			Product p = new Product();
			p.CategoryID = Convert.ToInt32(reader["CategoryID"]);
			p.Discontinued = Convert.ToBoolean(reader["Discontinued"]);
			p.ProductID  = Convert.ToInt32(reader["ProductID"]);
			p.ProductName  = Convert.ToString(reader["ProductName"]);
			p.QuantityPerUnit = Convert.ToString(reader["QuantityPerUnit"]);
			p.ReorderLevel = Convert.ToInt32(reader["ReorderLevel"]);
			p.SupplierID = Convert.ToInt32(reader["SupplierID"]);
			p.UnitPrice = Convert.ToDecimal(reader["UnitPrice"]);
			p.UnitsInStock = Convert.ToInt32(reader["UnitsInStock"]);
			p.UnitsOnOrder = Convert.ToInt32(reader["UnitsOnOrder"]);
			return p;
		}


		public event EventHandler Disposed;

		public ISite Site
		{
			get
			{
				return _site;
			}
			set
			{
				_site = value;
			}
		}

		/// <summary>
		/// IComponentの実装に必要です。
		/// </summary>
		public void Dispose()
		{
			OnDisposed(EventArgs.Empty);
		}

		/// <summary>
		/// IComponentの実装に必要です。
		/// </summary>
		protected virtual void OnDisposed(EventArgs e)
		{
			if (Disposed != null)
				Disposed(this, e);
		}
	}
}

File: DesktopSample/DataBinding/Section/IListBinding/VB.NET/IListBinding.DataLayer/ProductCollection.vb
Imports System
Imports System.ComponentModel
Imports System.Collections
Imports System.Data
Imports System.Data.SQLite

Public Class ProductCollection
	Inherits CollectionBase
	Implements IComponent

	Private _site As ISite

	Public Sub New()
		FillWithAllProducts()
		_site = Nothing
	End Sub 'New

	'<summary>
	' FillWithAllProducts - DBからのすべての製品のリストを作成するために使用
	'</summary>
	Public Sub FillWithAllProducts()
		'現在のリストをリセットする
		List.Clear()

		'開いている接続とデータを読み取る data
		Dim cn As SQLiteConnection = DataProvider.NewConnection
		Try
			cn.Open()
			Dim cmd As New SQLiteCommand("SELECT * FROM Products", cn)
			Dim reader As IDataReader = cmd.ExecuteReader()

			While reader.Read()
				List.Add(ProductFromDataReader(reader))
			End While
		Finally
			cn.Dispose()
		End Try
	End Sub

	'<summary>
	' ProductFromDataReader - リストのデータベースから新しい製品を作成するために使用
	'</summary>
	'<param name="reader">データを含む有効なDataReaderオブジェクト</param>
	'<returns>データ行から、新しく作成された製品固有のオブジェクト</returns>
	Friend Shared Function ProductFromDataReader(ByVal reader As IDataReader) As Product
		Dim p As New Product
		p.CategoryID = Convert.ToInt32(reader("CategoryID"))
		p.Discontinued = Convert.ToBoolean(reader("Discontinued"))
		p.ProductID = Convert.ToInt32(reader("ProductID"))
		p.ProductName = Convert.ToString(reader("ProductName"))
		p.QuantityPerUnit = Convert.ToString(reader("QuantityPerUnit"))
		p.ReorderLevel = Convert.ToInt32(reader("ReorderLevel"))
		p.SupplierID = Convert.ToInt32(reader("SupplierID"))
		p.UnitPrice = Convert.ToDecimal(reader("UnitPrice"))
		p.UnitsInStock = Convert.ToInt32(reader("UnitsInStock"))
		p.UnitsOnOrder = Convert.ToInt32(reader("UnitsOnOrder"))
		Return p
	End Function 'ProductFromDataReader

	Public Event Disposed As EventHandler Implements IComponent.Disposed

	'IComponentの実装に必要な
	Public Sub Dispose() Implements IComponent.Dispose
		'きれいにするものは何もありません。
		RaiseEvent Disposed(Me, EventArgs.Empty)
	End Sub

	'IComponentの実装に必要な
	Public Property Site() As ISite Implements IComponent.Site
		Get
			Return _site
		End Get
		Set(ByVal Value As ISite)
			_site = Value
		End Set
	End Property

End Class 'ProductCollection
